src/transformers/integrations/bitnet.py

```diff
@@ -0,0 +1,195 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from ..utils import is_accelerate_available, logging
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from accelerate import init_empty_weights
+if is_accelerate_available():
+    from accelerate import init_empty_weights
+
+logger = logging.get_logger(__name__)
+
+@torch.compile
+def unpack_weights(packed: torch.Tensor, bits: int = 2) -> torch.Tensor:
+    values_per_item = 8 // bits
+    packed_shape = packed.shape
+
+    if len(packed_shape) == 1:
+        original_row_dim = packed_shape[0] * values_per_item
+        unpacked_shape = (original_row_dim,)
+    else:
+        original_row_dim = packed_shape[0] * values_per_item
+        unpacked_shape = (original_row_dim, *packed_shape[1:])
+
+    unpacked = torch.zeros(unpacked_shape, device=packed.device, dtype=torch.uint8)
+
+    for i in range(values_per_item):
+        start = i * packed_shape[0]
+        end = start + packed_shape[0]
+        mask = (3 << (2 * i))
+        unpacked[start:end] = (packed & mask) >> (2 * i)
+
+    unpacked = unpacked.to(torch.bfloat16) - 1
+    return unpacked.to(torch.bfloat16)
+
+class BitLinear158(nn.Module):
+
+    def __init__(self, in_features: int, out_features: int, bias: bool = False, input_bits: int = 8,
+                 device=None, dtype=None, config=None):
+        super().__init__()
+        self.in_features = in_features
+        self.out_features = out_features
+        self.input_bits = input_bits
+        self.register_buffer(
+            "weight",
+            torch.zeros(
+                (out_features // 4, in_features),
+                dtype=torch.uint8,
+                device=device,
+            ),
+        )
+        self.register_buffer(
+            "weight_scale",
+            torch.ones(
+                (1),
+                dtype=torch.bfloat16,
+                device=device,
+            ),
+        )
+        if bias : 
+            self.register_buffer(
+                "bias",
+                torch.zeros(
+                    (self.out_features),
+                    dtype=torch.bfloat16,
+                    device=device
+                ))
+        else : 
+            self.bias = None
+        
+    @torch.compile
+    def activation_quant(self, x, num_bits=8):
+        Qn = -(2**(num_bits - 1))
+        Qp = 2**(num_bits - 1) - 1
+        s = Qp / x.abs().max(dim=-1, keepdim=True).values.clamp(min=1e-5)
+        result = (x * s).round().clamp(Qn, Qp)
+        return result.type(torch.int8), s
```

Comment:
Just a preference, both works 
```suggestion
        return result.to(torch.int8), s
```